src/Microsoft.Graph.Core/Requests/AuthenticationHandler.cs

```diff
@@ -0,0 +1,120 @@
+// ------------------------------------------------------------------------------
+
+namespace Microsoft.Graph
+{
+    using System.Net.Http;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using System.Net;
+    /// <summary>
+    /// A <see cref="DelegatingHandler"/> implementation using standard .NET libraries.
+    /// </summary>
+    public class AuthenticationHandler: DelegatingHandler
+    {
+        /// <summary>
+        /// MaxRetry property for 401's
+        /// </summary>
+        public int MaxRetry { get; set; } = 1;
+
+        /// <summary>
+        /// AuthenticationProvider property
+        /// </summary>
+        public IAuthenticationProvider AuthenticationProvider { get; set; }
+
+        /// <summary>
+        /// Construct a new <see cref="AuthenticationHandler"/>
+        /// </summary>
+        public AuthenticationHandler()
+        {
+
+        }
+
+        /// <summary>
+        /// Construct a new <see cref="AuthenticationHandler"/>
+        /// <param name="authenticationProvider">An authentication provider to pass to <see cref="AuthenticationHandler"/> for authenticating requests.</param>
+        /// </summary>
+        public AuthenticationHandler(IAuthenticationProvider authenticationProvider)
+        {
+            AuthenticationProvider = authenticationProvider;
+        }
+
+        /// <summary>
+        /// Construct a new <see cref="AuthenticationHandler"/>
+        /// </summary>
+        /// <param name="authenticationProvider">An authentication provider to pass to <see cref="AuthenticationHandler"/> for authenticating requests.</param>
+        /// <param name="innerHandler">A HTTP message handler to pass to the <see cref="AuthenticationHandler"/> for sending requests.</param>
+        public AuthenticationHandler(IAuthenticationProvider authenticationProvider, HttpMessageHandler innerHandler)
+        {
+            InnerHandler = innerHandler;
+            AuthenticationProvider = authenticationProvider;
+        }
+
+        /// <summary>
+        /// Checks HTTP response message status code if it's unauthorized (401) or not
+        /// </summary>
+        /// <param name="httpResponseMessage">The <see cref="HttpResponseMessage"/>to send.</param>
+        /// <returns></returns>
+        private bool IsUnauthorized(HttpResponseMessage httpResponseMessage)
+        {
+            return httpResponseMessage.StatusCode == HttpStatusCode.Unauthorized;
+        }
+
+        /// <summary>
+        /// Retry sending HTTP request
+        /// </summary>
+        /// <param name="httpResponseMessage">The <see cref="HttpResponseMessage"/>to send.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/>to send.</param>
+        /// <returns></returns>
+        private async Task<HttpResponseMessage> SendRetryAsync(HttpResponseMessage httpResponseMessage, CancellationToken cancellationToken)
```

Comment:
We will need to consider the following behavior where an Authorization header is not expected per a workload:
https://developer.microsoft.com/en-us/graph/docs/api-reference/v1.0/api/driveitem_createuploadsession#remarks
SendRetryAsync will cause an error if a Authorization header is provided in this scenario. Other than that, this PR looks good.

Reply:
this won't be an issue as we stap the Authorization header only for graph.microsoft.com issues.